sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamSqlEnv.java

```diff
@@ -73,109 +80,182 @@ public static BeamSqlEnv inMemory(TableProvider... tableProviders) {
+      functionSet.add(new SimpleEntry<>(functionName, UdfImpl.create(clazz, method)));
 
-  public Map<String, String> getPipelineOptions() {
-    return connection.getPipelineOptionsMap();
-  }
+      return this;
+    }
 
-  public String explain(String sqlString) throws ParseException {
-    try {
-      return RelOptUtil.toString(planner.convertToBeamRel(sqlString));
-    } catch (ValidationException | RelConversionException | SqlParseException e) {
-      throw new ParseException("Unable to parse statement", e);
+    /** Register a UDF function which can be used in SQL expression. */
+    public BeamSqlEnvBuilder registerUdf(String functionName, Class<? extends BeamSqlUdf> clazz) {
+      return registerUdf(functionName, clazz, BeamSqlUdf.UDF_METHOD);
+    }
+
+    public BeamSqlEnvBuilder registerUdf(String functionName, SerializableFunction sfn) {
+      return registerUdf(functionName, sfn.getClass(), "apply");
+    }
+
+    /**
+     * Register a UDAF function which can be used in GROUP-BY expression. See {@link
+     * org.apache.beam.sdk.transforms.Combine.CombineFn} on how to implement a UDAF.
+     */
+    public BeamSqlEnvBuilder registerUdaf(String functionName, Combine.CombineFn combineFn) {
+      functionSet.add(new SimpleEntry<>(functionName, new UdafImpl(combineFn)));
+      return this;
+    }
+
+    /** Load all UDF/UDAF from {@link UdfUdafProvider}. */
+    public BeamSqlEnvBuilder loadUdfUdafFromProvider() {
+      ServiceLoader.<UdfUdafProvider>load(UdfUdafProvider.class)
+          .forEach(
+              ins -> {
+                ins.getBeamSqlUdfs().forEach((udfName, udfClass) -> registerUdf(udfName, udfClass));
+                ins.getSerializableFunctionUdfs()
+                    .forEach((udfName, udfFn) -> registerUdf(udfName, udfFn));
+                ins.getUdafs().forEach((udafName, udafFn) -> registerUdaf(udafName, udafFn));
+              });
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder loadBeamBuiltinFunctions() {
+      for (BeamBuiltinFunctionProvider provider :
+          ServiceLoader.load(BeamBuiltinFunctionProvider.class)) {
+        registerBuiltinUdf(provider.getBuiltinMethods());
+      }
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder setQueryPlannerClassName(String name) {
+      queryPlannerClassName = name;
+      return this;
+    }
+
+    /**
+     * Build function to create an instance of BeamSqlEnv based on preset fields.
+     *
+     * @return BeamSqlEnv.
+     */
+    public BeamSqlEnv build() {
+      // This check is to retain backward compatible because most of BeamSqlEnv are initialized by
+      // withTableProvider API.
+      if (initialTableProvider == null) {
+        throw new RuntimeException("initialTableProvider must be set in BeamSqlEnvBuilder.");
+      }
+
+      JdbcConnection jdbcConnection = JdbcDriver.connect(initialTableProvider);
+
+      // set schema
```

Comment:
Extract into a function. The comment like `// set schema` means that there should be a function `setSchema()` instead.